### PR TITLE
add libdavs2 decoder  -hwrenx

### PR DESCRIPTION
**Pull request by hwrenx**

### CODE REVIEW

File: libavcodec/libdavs2.c [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    return 0;
+}
+
+static int DumpFrames(AVCodecContext *avctx, davs2_picture_t *pic, davs2_seq_info_t *headerset, AVFrame *frame)
+{
+    DAVS2Context *cad = avctx->priv_data;
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    if (!headerset)
+        return 0;
```
Comment (T-bagwell) [inline libavcodec/libdavs2.c:70]: What about use -EINVAL?

File: libavcodec/libdavs2.c [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    if (!pic || pic->ret_type == DAVS2_GOT_HEADER) {
+        avctx->width        = headerset->horizontal_size;
+        avctx->height       = headerset->vertical_size;
+        avctx->pix_fmt      = (headerset->output_bitdepth == 10 ? AV_PIX_FMT_YUV420P10 : AV_PIX_FMT_YUV420P);
```
Comment (T-bagwell) [inline libavcodec/libdavs2.c:75]: avctx->pix_fmt      = headerset->output_bitdepth == 10 ? AV_PIX_FMT_YUV420P10 : AV_PIX_FMT_YUV420P;

File: libavcodec/libdavs2.c [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+        AVRational r = av_d2q(headerset->frame_rate,4096);
+        avctx->framerate.num = r.num;
+        avctx->framerate.den = r.den;
+        return 0;
+    }
+
+    int bytes_per_sample = pic->bytes_per_sample;
+
+    int i;
```
Comment (T-bagwell) [inline libavcodec/libdavs2.c:85]: move to start of this function, same with bytes_per_sample

File: libavcodec/libdavs2.c [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    cad->frame = frame;
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    if (!buf_size) {
+        cad->packet.data = buf_ptr;
+        cad->packet.len  = buf_size;
+        cad->packet.pts  = avpkt->pts;
+        cad->packet.dts  = avpkt->dts;
+
+        for (;;) {
```
Comment (T-bagwell) [inline libavcodec/libdavs2.c:140]: Why don't use while instead for(;;)

File: libavcodec/libdavs2.c [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+            if (cad->out_frame.ret_type != DAVS2_DEFAULT) {
+                *got_frame = DumpFrames(avctx, &cad->out_frame, &cad->headerset, frame);
+                davs2_decoder_frame_unref(cad->decoder, &cad->out_frame);
+            }
+            if (*got_frame == 1)
+                break;
+        }
+        return 0;
+    } else {
+        for (; buf_size > 0;) {
```
Comment (T-bagwell) [inline libavcodec/libdavs2.c:155]: why don't use while(buf_size > 0)?

File: libavcodec/libdavs2.c [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+                    av_log(avctx, AV_LOG_VERBOSE, "decoder destroyed. 0x%p; frames %d\n", cad->decoder, cad->decoded_frames);
+                    cad->decoder = NULL;
+                }
+                return AVERROR(EINVAL);
+            }
+
+            buf_ptr     += len;
+            buf_size    -= len;
+
+            if (*got_frame == 1)
```
Comment (T-bagwell) [inline libavcodec/libdavs2.c:184]: Is this a bool? just *got_frame or !*got_frame?  it the DumpFrames is a bool type?
and the API name "DumpFrames" is not ffmpeg code style , what about use davs_dump_frames?

File: libavcodec/libdavs2.c (hunk 1 above)
Comment (hwrenx) [reply libavcodec/libdavs2.c:70]: AVERROR(EINVAL)? better indeed.

File: libavcodec/libdavs2.c [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+#include "libavutil/common.h"
+#include "libavutil/avutil.h"
+#include "avcodec.h"
+#include "libavutil/imgutils.h"
+#include "internal.h"
+
+#include <davs2.h>
+
+typedef struct DAVS2Context {
+    AVCodecContext *avctx;
```
Comment (mypopydev) [inline libavcodec/libdavs2.c:35]: Usually we put AVClass as the first field, I don't think need to use avctx in a private context.

File: libavcodec/libdavs2.c [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    .type           = AVMEDIA_TYPE_VIDEO,
+    .id             = AV_CODEC_ID_AVS2,
+    .priv_data_size = sizeof(DAVS2Context),
+    .init           = ff_davs2_init,
+    .close          = ff_davs2_end,
+    .decode         = davs2_decode_frame,
+    .capabilities   =  AV_CODEC_CAP_DELAY,//AV_CODEC_CAP_DR1 |
+    .pix_fmts       = (const enum AVPixelFormat[]) { AV_PIX_FMT_YUV420P, AV_PIX_FMT_YUV420P10,
+                                                     AV_PIX_FMT_NONE },
+};
```
Comment (mypopydev) [inline libavcodec/libdavs2.c:206]: Missing wrapped_name field when uses 3rd library, same as the encoder

File: libavcodec/libdavs2.c [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+static av_cold ff_davs2_init(AVCodecContext *avctx)
+{
+    DAVS2Context *cad = avctx->priv_data;
+
+    /* init the decoder */
+    cad->param.threads      = 0;
+    cad->param.i_info_level = 0;
+    cad->decoder = davs2_decoder_open(&cad->param);
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
```
Comment (mypopydev) [inline libavcodec/libdavs2.c:58]: Why ?

File: libavcodec/libdavs2.c [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+    davs2_seq_info_t headerset;  // output data, sequence header
+
+}DAVS2Context;
+
+static av_cold ff_davs2_init(AVCodecContext *avctx)
+{
+    DAVS2Context *cad = avctx->priv_data;
+
+    /* init the decoder */
+    cad->param.threads      = 0;
```
Comment (mypopydev) [inline libavcodec/libdavs2.c:55]: Can we use an option to enable multiple-thread?

File: libavcodec/libdavs2.c [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    av_log(avctx, AV_LOG_VERBOSE, "decoder created. 0x%p\n", cad->decoder);
+    return 0;
+}
+
+static int DumpFrames(AVCodecContext *avctx, davs2_picture_t *pic, davs2_seq_info_t *headerset, AVFrame *frame)
+{
+    DAVS2Context *cad = avctx->priv_data;
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
```
Comment (mypopydev) [inline libavcodec/libdavs2.c:67]: Same as above

File: libavcodec/libdavs2.c [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    uint8_t *buf_ptr = avpkt->data;
+    AVFrame *frame = data;
+    int ret = 0;
+
+    *got_frame = 0;
+    cad->frame = frame;
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    if (!buf_size) {
+        cad->packet.data = buf_ptr;
```
Comment (mypopydev) [inline libavcodec/libdavs2.c:135]: avpkt->data and avpkt->size? We don't need another variable in this case

File: libavcodec/libdavs2.c (hunk 4 above)
Comment (mypopydev) [inline libavcodec/libdavs2.c:140]: Maybe we can use new API decode/receive_frame callback to decouple input packet/the output frame, this is the refer link: https://blogs.gentoo.org/lu_zero/2016/03/29/new-avcodec-api/

File: libavcodec/libdavs2.c (hunk 9 above)
Comment (hwrenx) [reply libavcodec/libdavs2.c:58]: we won't send complete frames to our decoder, so we change the flag. If not, we may get wrong info about the packet.

File: libavcodec/libdavs2.c (hunk 10 above)
Comment (hwrenx) [reply libavcodec/libdavs2.c:55]: maybe use avctx->thread_count?

File: libavcodec/libdavs2.c (hunk 12 above)
Comment (hwrenx) [reply libavcodec/libdavs2.c:135]: we need this two pointers when we send the next part of data.

File: libavcodec/libdavs2.c (hunk 4 above)
Comment (hwrenx) [reply libavcodec/libdavs2.c:140]: we may discuss it later.

**Pull request closed without merging**